src/rgw/rgw_notify.cc

```diff
@@ -493,9 +493,18 @@ class Manager : public DoutPrefixProvider {
       // start the worker threads to do the actual queue processing
       const std::string WORKER_THREAD_NAME = "notif-worker";
       for (auto worker_id = 0U; worker_id < worker_count; ++worker_id) {
-        workers.emplace_back([this]() noexcept { io_context.run(); });
+        workers.emplace_back([this]() noexcept {
+          while (true) {
+            try {
+              io_context.run(); 
+              break; // exited notmally
+            } catch (const std::exception& err) {
+              ldpp_dout(this, 10) << "Notification worker failed with error: " << err.what() << dendl;
+            }
```

Comment:
ignoring exceptions here does keep the background threads running, but it doesn't actually recover if the exception was thrown in `process_queues()`. i'd argue it's better to crash and restart on an unexpected exception

Reply:
this is the backtrace for the exception (when not caught):
```
#0  0x00007f3e444159d5 in raise () from /lib64/libc.so.6
#1  0x00007f3e443fe8a4 in abort () from /lib64/libc.so.6
#2  0x00007f3e447c6926 in __gnu_cxx::__verbose_terminate_handler() [clone .cold] () from /lib64/libstdc++.so.6
#3  0x00007f3e447d21ac in __cxxabiv1::__terminate(void (*)()) () from /lib64/libstdc++.so.6
#4  0x00007f3e447d1239 in __cxa_call_terminate () from /lib64/libstdc++.so.6
#5  0x00007f3e447d1b81 in __gxx_personality_v0 () from /lib64/libstdc++.so.6
#6  0x00007f3e445d5c2f in _Unwind_RaiseException_Phase2 () from /lib64/libgcc_s.so.1
#7  0x00007f3e445d665e in _Unwind_Resume () from /lib64/libgcc_s.so.1
#8  0x0000564497356f9b in boost::asio::detail::scheduler::run (this=0x564498b9d000, ec=...) at /root/projects/ceph/build/boost/include/boost/asio/detail/impl/scheduler.ipp:193
#9  0x00007f3e516a4ab8 in boost::asio::io_context::run (this=0x564498bc7f18) at /root/projects/ceph/build/boost/include/boost/asio/impl/io_context.ipp:63
#10 0x00007f3e521b50bd in rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal::RGWRadosStore*)::{lambda()#2}::o
perator()() const (__closure=0x564498ba6a98) at /root/projects/ceph/src/rgw/rgw_notify.cc:497
#11 0x00007f3e521c80de in std::__invoke_impl<void, rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal::RGWRado
sStore*)::{lambda()#2}>(std::__invoke_other, rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal::RGWRadosStore
*)::{lambda()#2}&&) (__f=...) at /usr/include/c++/10/bits/invoke.h:60
#12 0x00007f3e521c7fb6 in std::__invoke<rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal::RGWRadosStore*)::{
lambda()#2}>(rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal::RGWRadosStore*)::{lambda()#2}&&) (__fn=...) a
t /usr/include/c++/10/bits/invoke.h:95
#13 0x00007f3e521c7e86 in std::thread::_Invoker<std::tuple<rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal:
:RGWRadosStore*)::{lambda()#2}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x564498ba6a98) at /usr/include/c++/10/thread:264
#14 0x00007f3e521c7898 in std::thread::_Invoker<std::tuple<rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, rgw::sal:
:RGWRadosStore*)::{lambda()#2}> >::operator()() (this=0x564498ba6a98) at /usr/include/c++/10/thread:271
#15 0x00007f3e521c7406 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<rgw::notify::Manager::Manager(ceph::common::CephContext*, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int
, unsigned int, rgw::sal::RGWRadosStore*)::{lambda()#2}> > >::_M_run() (this=0x564498ba6a90) at /usr/include/c++/10/thread:215
#16 0x00007f3e447fe5f4 in execute_native_thread_routine () from /lib64/libstdc++.so.6
#17 0x00007f3e445ac3f9 in start_thread () from /lib64/libpthread.so.0
#18 0x00007f3e444d9903 in clone () from /lib64/libc.so.6
```

Reply:
since coroutines run on their own stack, there's no way for outside code to catch their exceptions. so the spawn library catches exceptions internally, and rethrows them after returning to the calling thread in io_context.run()

Reply:
actual backtrace: https://paste.sh/HbPH_uvB#oBb-24wzX6sQwAF9EX5YftX8

Reply:
found the actual issue: queues garbage collector moved out of the loop to prevent a dangling reference inside the coroutine.

Reply:
ok great!

good idea to log the exception - you can drop the while() and break parts though